automation/utilities/rediswq.py

```diff
@@ -58,24 +74,127 @@ def empty(self):
         """
         return self._main_qsize() == 0 and self._processing_qsize() == 0
 
+    def _maybe_renew_job(self, job):
+        """Transactionally move job from the processing to the work queue.
+
+        A job will not be renewed if it appears that another worker is
+        changing the processing queue or it the job has exceeded the
+        maximum number of retries.
+        """
+        self._logger.debug(
+            "Lease expired for job %s. Attempting to move from processing "
+            "queue to work queue...[session %s]" % (job, self.sessionID())
+        )
+        # A pipeline executes a set of commands as a transaction, but
+        # does not support rollback in the event any individual
+        # command fails.
+        pipe = self._db.pipeline(transaction=True)
+
+        # First, watch the processing queue and retry count hash map for
+        # changes. The pipeline transaction won't execute if changes occur
+        # after this point. This allows us to prevent duplicate updates from
+        # different workers.
+        pipe.watch(self._processing_q_key)
+        pipe.watch(self._retry_hash_map_key)
+
+        # Next, ensure that the job is still in the processing queue.
+        # The processing queue is relatively small, so this shouldn't
+        # be an expensive operation.
+        if job not in self._db.lrange(self._processing_q_key, 0, -1):
+            pipe.reset()
+            self._logger.debug(
+                "Job %s no longer in the processing queue, there is no need "
+                "to renew..." % job
+            )
+            return
+
+        # Next, check that we haven't exceeded the retry limit for this job
+        # If we have, remove the job from the queue entirely and don't add it
+        # back to the main work queue
```

Comment:
Do we get a crawl_history record here indicating that it failed more than n times?

Reply:
No, but it's something I can add.

Reply:
I think it would be very useful. Maybe even adding a record to crawl_history before processing starts. Anything to make it easier to understand what happened with each site crawl attempt :)

Reply:
Agree, it is out of scope for this PR, which is good to merge now :)